---
title: "Consensus Visualization"
editor: visual
jupyter: python3
---

## Creating coverage plots

[weeSAM](https://github.com/centre-for-virus-research/weeSAM) can be used for generating coverage plots and providing statistics based on the bam file.

Make sure you are in the right path

``` bash
cd /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Viro_Run_0001/barcode01/
```

::: callout-warning
## Warning

Remember to update the run paths and barcodes for each sample!
:::

Now just copy the command below:

``` bash
weeSAM --bam all_reads.bam --html coverage_plot
```

![Example](coverage.jpg)

In the newly created folder **"coverage_plot_html_results"**, click on "figures" then "all_reads_figures" until you find the html report.
