---
title: "Generating a Consensus Sequence"
editor: visual
jupyter: python3
---

After confirming read identification (via taxonomic classification) or contig (via assembly), a consensus sequence can be generated using the closest matching reference available on NCBI.

Make sure you are in the right environment (you only need to change environments if you have performed assembly in the previous step)

``` bash
conda activate nanopore_diagnostics
```

::: callout-tip
## Tip

If your target organism is frequently detected and exhibits substantial genetic diversity or multiple genotypes (e.g., enteroviruses or human herpesvirus C), you may consider creating an in-house reference database to facilitate identification of the best sequence match. This database should be carefully curated to ensure accuracy and relevance.
:::

## Create a "Reference" folder

This will serve as a path for all your future references. This is really important as it means your path will always stay the same and you only need to change the reference name (i.e the fasta file name). First check if this folder has already been made, if not create a Reference folder. 

``` bash
cd /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/
ls
```

## Download the most appropriate reference fasta file

You can find this by blasting your longest read (Kraken2) or contig (assembly) on NCBI. Pay close attention to the reference size, coverage and percentage identity.

![Example](Ref.jpg)

Rename the reference fasta to avoid any spaces or unsuitable characters. Its best to keep it short and to the point - accession number and target. For example, "HQ647172_Enterovirus_A71.fasta"

## Make sure you are in the right path

``` bash
cd /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Viro_Run_0001/barcode01/
```

## Consensus generation

The script works by reading the mapping file position-by-position and counting the highest nucleotide at each position. If a deletion (compared to the original reference) has the highest count, this will be taken into consideration during consensus generation.

This step involves a few commands. Lets break it up!

::: callout-warning
## Warning

Remember to update the reference path
:::

### Read alignment and indexing

MiniMap2 is used for mapping our selected reference to our trimmed reads.

``` bash
minimap2 -Y -t 32 -x map-ont -a /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Reference/HQ647172_Enterovirus_A71.fasta all_reads_QC_hg19.fastq 2> /dev/null | samtools view -bF 4 - | samtools sort -@ 16 - > all_reads.bam
samtools index all_reads.bam
```

ðŸ”¹<strong style="color:darkblue">-Y</strong> â€“ Use soft clipping for supplementary alignments (preserves soft-clipped bases at the ends).\
ðŸ”¹<strong style="color:darkblue">-t 32</strong> â€“ Use 32 threads for faster processing.\
ðŸ”¹<strong style="color:darkblue">-b</strong> â€“ Output BAM format.\
ðŸ”¹<strong style="color:darkblue">-F 4</strong> â€“ Filter unmapped reads (i.e., include only mapped reads).\
ðŸ”¹<strong style="color:darkblue">-\@ 16</strong> â€“ Use 16 threads for sorting.

### Generating depth per position and calling varients

``` bash
samtools mpileup -a -A -Q 0 -d 0 -f /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Reference/HQ647172_Enterovirus_A71.fasta all_reads.bam | cut -f 2,3,4 --output-delimiter=',' > all_reads.depth
/mnt/scratch2/other_scripts/NAW_helperscripts/bam2vcf.py -c 8 -d 1 -af 0.1 -r /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Reference/HQ647172_Enterovirus_A71.fasta -b all_reads.bam -o all_reads.vcf
/mnt/scratch2/other_scripts/NAW_helperscripts/filtervcf.py -i all_reads.vcf -o all_reads_filtered.vcf
```

ðŸ”¹<strong style="color:darkblue">-a</strong> â€“ Include all positions, even those with 0 coverage.\
ðŸ”¹<strong style="color:darkblue">-A</strong> â€“ Include anomalous read pairs (not excluded by default).\
ðŸ”¹<strong style="color:darkblue">-Q</strong> â€“ Minimum base quality threshold is 0 (include all reads).\
ðŸ”¹<strong style="color:darkblue">-d</strong> â€“ Disable maximum read depth limit (default is 8000).\
ðŸ”¹<strong style="color:darkblue">-f</strong> â€“ Reference genome.\
ðŸ”¹<strong style="color:darkblue">-c</strong> â€“ Minimum base quality (PHRED) score for a base to be considered.\
ðŸ”¹<strong style="color:darkblue">-d</strong> â€“ Minimum depth to call a variant.\
ðŸ”¹<strong style="color:darkblue">-af</strong> â€“ Minimum allele frequency (e.g., 0.1 = 10%).\
ðŸ”¹<strong style="color:darkblue">-r</strong> â€“ Reference genome.\
ðŸ”¹<strong style="color:darkblue">-b</strong> â€“ BAM file.\
ðŸ”¹<strong style="color:darkblue">-o</strong> â€“ Output VCF file.

::: callout-important
## Important

Depth (-d) is currently set at 1. For initial analysis this is fine, however its a good idea to set at least 3 for better confidence. If you are looking specifically at SNPs or antiviral resistance, depending on the target, consider changing to "30" or "100".
:::

### Generating a consensus sequence

``` bash
/mnt/scratch2/other_scripts/NAW_helperscripts/vcf2consensus.py -v all_reads_filtered.vcf -d all_reads.depth -r /mnt/viro0002-data/sequencedata/processed/Diagnostics_metagenomics/Reference/HQ647172_Enterovirus_A71.fasta -o all_reads_consensus.fasta -dp 1 -n "Consensus_1x"
samtools fastq all_reads.bam > all_reads_mapped_readcount.fastq
```

ðŸ”¹<strong style="color:darkblue">-v</strong> â€“ Input VCF file.\
ðŸ”¹<strong style="color:darkblue">-d</strong> â€“ Depth file (from mpileup) used to decide whether to mask low coverage sites.\
ðŸ”¹<strong style="color:darkblue">-r</strong> â€“ Reference sequence.\
ðŸ”¹<strong style="color:darkblue">-o</strong> â€“ Output consensus FASTA file.\
ðŸ”¹<strong style="color:darkblue">-dp</strong> â€“ Minimum depth required to call a base; otherwise, likely masked as "N".\
ðŸ”¹<strong style="color:darkblue">-n</strong> â€“ Name for the FASTA header.

::: callout-important
## Important

Depth (-dp) is currently set at 1. For initial analysis this is fine, however its a good idea to set at least 3 for better confidence. If you are looking specifically at snps or antiviral resistance, depending on the target, consider changing to "30" or "100". Remember to update the fasta name (-n).
:::

### Read mapping statistics

Next, we can generate read statistics using [Fastp](https://github.com/OpenGene/fastp), a flexible tool for quality control and processing of FASTQ data. Fastp can be used to remove low-quality reads and short sequences. Pay close attention to the total number of reads (post-QC), the total number of nucleotides (post-QC), and the average read length. Record these values for each sample in the Excel results sheet.

``` bash
fastp -i all_reads_mapped_readcount.fastq -o all_reads_mapped_readcount_QC.fastq -j /dev/null -h all_reads_mapped_readcount_QC.html --disable_trim_poly_g --disable_adapter_trimming --qualified_quality_phred 8 --unqualified_percent_limit 50 --length_required 150 -w 16
```

ðŸ”¹<strong style="color:darkblue">-i all_reads_mapped.fastq</strong> â€“ Input FASTQ file which has just been mapped.\
ðŸ”¹<strong style="color:darkblue">-o all_reads_mapped_readcount_QC.fastq</strong> â€“ Output FASTQ file; this will be your filtered mapped reads.\
ðŸ”¹<strong style="color:darkblue">-j /dev/null</strong> â€“ Disables the JSON report as it's not required.\
ðŸ”¹<strong style="color:darkblue">-h all_reads_QC.html</strong> â€“ Produces a visual quality report used for read statistics.\
ðŸ”¹<strong style="color:darkblue">--disable_trim_poly_g</strong> â€“ Disables this trimming feature (more relevant to Illumina reads).\
ðŸ”¹<strong style="color:darkblue">--disable_adapter_trimming</strong> â€“ Disables adapter trimming (assumes itâ€™s already handled or unnecessary).\
ðŸ”¹<strong style="color:darkblue">--qualified_quality_phred 8</strong> â€“ A base is "qualified" if Phred â‰¥ 8.\
ðŸ”¹<strong style="color:darkblue">--unqualified_percent_limit 50</strong> â€“ Max 50% of bases can be low quality; reads exceeding this are discarded.\
ðŸ”¹<strong style="color:darkblue">--length_required 150</strong> â€“ Discards reads shorter than 150 bases after filtering.\
ðŸ”¹<strong style="color:darkblue">-w 16</strong> â€“ Sets the number of threads to 16.

### Quality control

You now have your initial consensus sequence. To increase your confidence in your sequence, always blast your consensus on NCBI. Using the top (or most appropriate reference), perform mapping again, this time using the updated reference.

Open the bam file (all_reads.bam) using [UGENE](https://ugene.net/) and check the coverage pattern. Ugene can be downloaded from the software center. Check if there is any coverage bias. You can also use this file to compare to the consensus and check for allele frequency.

![Example](bam.jpg)

### Typing

After consensus generation, you can create coverage plots and trees (shown in the next few chapters).

Depending on the target, you can also check out some typing tools. For example, [Genome Detective](https://www.genomedetective.com/app/typingtool/virus/) will automatically perform a subtyping analysis for the following supported viruses:

![Examples](GD_typing.jpg)
